### PR TITLE
Implement MediaStreamSourceNode

### PR DESCRIPTION
**Pull request by Manishearth**
<s>There's a gst-rs bug that causes this to spew some stuff to the console right now.</s> Introduced a hack as suggested by Sebastian.

Also I can't get audio capture to work but it *seems* like audio capture might be broken in general on my Linux machine.

Fixes https://github.com/servo/media/issues/343

### CODE REVIEW

**Comment by Manishearth** [comment]
@ferjm would you be able to run examples/audioinput_stream.rs and confirm that it works?

**Comment by sdroege** [comment]
> There's a gst-rs bug

See https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/-/issues/779 and https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/-/merge_requests/1363. The latter would solve that problem for you and not require a capsfilter anymore (but will have the same effect overall).

**Comment by Manishearth** [comment]
Ah, the caps were bubbling out. I see.

File: examples/Cargo.toml [hunk 1]
```diff
@@ -37,6 +38,14 @@ webrtc = [ "rand", "serde", "serde_derive", "serde_json", "websocket" ]
 name = "dummy"
 path = "dummy.rs"
 
+[[bin]]
+name = "audioinput_stream"
+path = "audioinput_stream.rs"
+
+[[bin]]
+name = "videoinput_stream"
+path = "videoinput_stream.rs"
```
Comment (ferjm) [inline examples/Cargo.toml:47]: This file is not included in the PR

**Comment by ferjm** [comment]
> @ferjm would you be able to run examples/audioinput_stream.rs and confirm that it works?

(I replied on Matrix, but I reply here as well for posterity.)

I get no sound and a few warnings on the GST_DEBUG output

```
     Running `target/debug/audioinput_stream`
0:00:00.110808000 27465 0x7fcfc180a240 FIXME                default gstutils.c:3981:gchar *gst_pad_create_stream_id_internal(GstPad *, GstElement *, const gchar *):<osxaudiosrc0:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
0:00:00.123759000 27465 0x7fcfc180a1e0 WARN         audio-resampler audio-resampler.c:274:void convert_taps_gint16_c(gdouble *, gpointer, gdouble, gint): can't find exact taps
0:00:00.811785000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.023123931 < -+0:00:00.020000000
0:00:00.854511000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.020598358 < -+0:00:00.020000000
0:00:00.865092000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.022513348 < -+0:00:00.020000000
0:00:00.918423000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.021629278 < -+0:00:00.020000000
0:00:00.939799000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.020576218 < -+0:00:00.020000000
0:00:01.035745000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.021655065 < -+0:00:00.020000000
0:00:01.121107000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.021232611 < -+0:00:00.020000000
0:00:01.334394000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.020615676 < -+0:00:00.020000000
0:00:02.145076000 27465 0x7fcfc180a180 WARN           audiobasesink gstaudiobasesink.c:1491:void gst_audio_base_sink_skew_slaving(GstAudioBaseSink *, GstClockTime, GstClockTime, GstClockTime *, GstClockTime *):<autoaudiosink0-actual-sink-osxaudio> correct clock skew -0:00:00.020090777 < -+0:00:00.020000000
```

**Comment by Manishearth** [comment]
Seems like audioinput streams are broken then.

File: examples/audioinput_stream.rs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+extern crate servo_media;
+extern crate servo_media_auto;
+
+use servo_media::audio::node::{AudioNodeInit, AudioNodeMessage, AudioScheduledSourceNodeMessage};
```
Comment (ferjm) [inline examples/audioinput_stream.rs:4]: These are not used

File: examples/audioinput_stream.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+use servo_media::{ClientContextId, ServoMedia};
```
Comment (ferjm) [inline examples/audioinput_stream.rs:5]: `ClientContextId` is not used

File: examples/videoinput_stream.rs [hunk 4]
```diff
@@ -0,0 +1,25 @@
+extern crate servo_media;
+extern crate servo_media_auto;
+
+use servo_media::audio::node::{AudioNodeInit, AudioNodeMessage, AudioScheduledSourceNodeMessage};
+use servo_media::{ClientContextId, ServoMedia};
```
Comment (ferjm) [inline examples/videoinput_stream.rs:5]: Same unused imports as `audioinput_stream`

**Comment by Manishearth** [comment]
@bors-servo r=ferjm

CI doesn't have input elements

**Comment by Manishearth** [comment]
Er, input sources, the elements exist.

**Pull request merged**